x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/TextExpansionQueryBuilder.java

```diff
@@ -0,0 +1,266 @@
+            return new TextExpansionQueryBuilder(this, query);
+        }
+
+        InferModelAction.Request inferRequest = InferModelAction.Request.forTextInput(
+            modelId,
+            SlimConfigUpdate.EMPTY_UPDATE,
+            List.of(modelText)
+        );
+
+        SetOnce<SlimResults> slimResultsSupplier = new SetOnce<>();
+        queryRewriteContext.registerAsyncAction((client, listener) -> {
+            executeAsyncWithOrigin(client, ML_ORIGIN, InferModelAction.INSTANCE, inferRequest, ActionListener.wrap(inferenceResponse -> {
+
+                if (inferenceResponse.getInferenceResults().isEmpty()) {
+                    listener.onFailure(new IllegalStateException("inference response contain no results"));
+                    return;
+                }
+
+                if (inferenceResponse.getInferenceResults().get(0)instanceof SlimResults slimResults) {
+                    slimResultsSupplier.set(slimResults);
+                    listener.onResponse(null);
+                } else if (inferenceResponse.getInferenceResults().get(0)instanceof WarningInferenceResults warning) {
+                    listener.onFailure(new IllegalStateException(warning.getWarning()));
+                } else {
+                    listener.onFailure(
+                        new IllegalStateException(
+                            "expected a result of type ["
+                                + SlimResults.NAME
+                                + "] received ["
+                                + inferenceResponse.getInferenceResults().get(0).getWriteableName()
+                                + "]. Is ["
+                                + modelId
+                                + "] a compatible model?"
+                        )
+                    );
+                }
+            }, listener::onFailure));
+        });
+
+        return new TextExpansionQueryBuilder(this, slimResultsSupplier);
+    }
+
+    static BoolQueryBuilder weightedTokensToQuery(String fieldName, SlimResults slimResults, QueryRewriteContext queryRewriteContext)
+        throws IOException {
+        var boolQuery = QueryBuilders.boolQuery();
+        for (var weightedToken : slimResults.getWeightedTokens()) {
+            boolQuery.should(
+                QueryBuilders.termQuery(fieldName, Integer.toString(weightedToken.token()))
+                    .boost(weightedToken.weight())
+                    .rewrite(queryRewriteContext)
+            );
+        }
+        boolQuery.minimumShouldMatch(1);
+        boolQuery.rewrite(queryRewriteContext);
+        return boolQuery;
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        if (termsQuery == null) {
+            throw new IllegalStateException("terms query must not be null, missing rewrite?");
+        }
+        return termsQuery.toQuery(context);
+    }
```

Comment:
I think that this should simply `throw` in all cases and rely on `rewrite` to return a `Bool#terms`